util/client/ngmclient/client.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2021 PingCAP, Inc. Licensed under Apache-2.0.
+
+package ngmclient
+
+import (
+	"context"
+	"fmt"
+	"net/http/httputil"
+	"net/url"
+	"sync/atomic"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/joomcode/errorx"
+	"go.etcd.io/etcd/clientv3"
+	"go.uber.org/fx"
+	"golang.org/x/sync/singleflight"
+
+	"github.com/pingcap/tidb-dashboard/pkg/utils/topology"
+	"github.com/pingcap/tidb-dashboard/util/topo"
+)
+
+var (
+	NgmErrNS                 = errorx.NewNamespace("error.api.ng_monitoring")
+	ErrNgMonitoringNotDeploy = NgmErrNS.NewType("not_deploy")
+	ErrNgMonitoringNotStart  = NgmErrNS.NewType("not_start")
+)
+
+const (
+	ngMonitoringCacheTTL = time.Second * 5
+)
+
+type ngMonitoringAddrCacheEntity struct {
+	address string
+	err     error
+	cacheAt time.Time
+}
```

Comment:
Seems that the way you use ngm is a good case of adopting a TopologyProvider in https://github.com/pingcap/tidb-dashboard/pull/1098.

Think about:

```go
type TargetTopoProvider interface {
    GetNgm(ctx context.Context) ([]topo.NgmInfo, error)
}

func NewProxy(target TargetTopoProvider) {
   ...
}

// usage:

ngmclient.NewProxy(myCachedTopologyProvider)
// the developer is free to assemble the proxy using a non-cached topology, 
// or own ngm topo provider! So this makes it useful in more cases.
```

Reply:
Yeah, it would be better to open a new task after Top SQL PR is finished.

Reply:
Let me bring back the ngm proxy impl.